### PR TITLE
Support generating bootstrapping keys at the server side. 

### PR DESCRIPTION
**Pull request by Guy-Moshkowich**
*Issue #22  

*Description of changes:*
This PR enable generating bootstrapping keys at the server side. Current API, requires the secret key which is not available in the server side. The new implementation require the generation of bootstrapping switching key pair and their marshaling.

1) Add support for un/marshal SwitchingKey 
2) Generate bootstrapping keys without secret key based on Lattigo v4.1.0 implementation https://github.com/tuneinsight/lattigo/blob/adf762375670e412bab261cd7ffff9ca03777ad5/ckks/bootstrapping/bootstrapper.go#L78
3) Enable generating conjugate rotation. Lattigo original API already supports this feature but the C++ wrapper used to call it with fixed "false" value. This is useful for generating bootstrapping keys manually.
4) Deprecated a function that accept unused parameters and fix bootstrapping example to use the fixed function

All the C++ wrapper examples are running successfully with this PR changes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/gowrapper/ckks/keygen.go [hunk 1]
```diff
@@ -336,7 +340,7 @@ func lattigo_setRotKeysForEvaluationKey(evalKeyHandle Handle5, rotKeysHandle Han
 }
 
 //export lattigo_genBootstrappingKey
-func lattigo_genBootstrappingKey(keygenHandle Handle5, paramHandle Handle5, btpParamsHandle Handle5, skHandle Handle5, relinKeyHandle Handle5, rotKeyHandle Handle5) Handle5 {
+func lattigo_genBootstrappingKey(keygenHandle Handle5, paramHandle Handle5, btpParamsHandle Handle5, skHandle Handle5) Handle5 {
```
Comment (crockeea) [inline src/gowrapper/ckks/keygen.go:343]: Nice catch.

File: src/latticpp/ckks/keygen.cpp [hunk 2]
```diff
@@ -139,8 +143,22 @@ namespace latticpp {
                                          rotKeys.getRawHandle());
     }
 
-    BootstrappingKey genBootstrappingKey(const KeyGenerator &keygen, const Parameters &params, const BootstrappingParameters &bootParams, const SecretKey &sk, const RelinearizationKey &relinKey, const RotationKeys &rotKeys) {
-        return BootstrappingKey(lattigo_genBootstrappingKey(keygen.getRawHandle(), params.getRawHandle(), bootParams.getRawHandle(), sk.getRawHandle(), relinKey.getRawHandle(), rotKeys.getRawHandle()));
+    // DEPRACATED
+    BootstrappingKey genBootstrappingKey(const KeyGenerator &keygen, const Parameters &params, const BootstrappingParameters &bootParams, const SecretKey &sk, const RelinearizationKey & /*relinKey*/, const RotationKeys &/*rotKeys*/) {
+        return genBootstrappingKey(keygen, params, bootParams, sk);
+    }
```
Comment (crockeea) [inline src/latticpp/ckks/keygen.cpp:149]: I'm not terribly concerned with backwards compatibility; I suggest we just kill this function.
```suggestion
```

File: src/latticpp/ckks/keygen.h [hunk 3]
```diff
@@ -63,6 +68,8 @@ namespace latticpp {
 
     RotationKeys genRotationKeysForRotations(const KeyGenerator &keygen, const SecretKey &sk, std::vector<int> shifts);
 
+    RotationKeys genRotationKeysForRotations(const KeyGenerator &keygen, const SecretKey &sk, std::vector<int> shifts, int includeConjugate);
```
Comment (crockeea) [inline src/latticpp/ckks/keygen.h:71]: Nit: We could just keep this API and set the default argument for `includeConjugate` to `false`.

File: src/latticpp/ckks/keygen.h [hunk 4]
```diff
@@ -75,8 +82,15 @@ namespace latticpp {
     void setRotKeysForEvaluationKey(const EvaluationKey &evalKey,
                                     const RotationKeys &rotKeys);
 
+    // DEPRACATED: use genBootstrappingKey with the same list of parameters excluding RelinearizationKey, RotationKeys
```
Comment (crockeea) [inline src/latticpp/ckks/keygen.h:85]:
```suggestion
```

**Review by crockeea — CHANGES_REQUESTED** [review]
Thank you so much for this contribution!

**Comment by Guy-Moshkowich** [comment]
Thank you for taking the time to review and for the good comments. I've fixed all of them.

**Review by crockeea — APPROVED** [review]

**Pull request merged**